source/lac/scalapack.cc

```diff
@@ -323,117 +323,298 @@ void ScaLAPACKMatrix<NumberType>::invert()
+                 &vl, &vu, &il, &iu, &abstol, &m, &nz, &ev[0], &orfac,
+                 eigenvectors_loc, &eigenvectors->submatrix_row, &eigenvectors->submatrix_column, eigenvectors->descriptor,
+                 &work[0], &lwork, &iwork[0], &liwork, &ifail[0], &iclustr[0], &gap[0], &info);
+        }
       lwork=work[0];
       work.resize (lwork);
 
-      psyev(&jobz, &uplo, &n_rows, A_loc, &submatrix_row, &submatrix_column, descriptor, &ev[0],
-            Z_loc, &Z.submatrix_row, &Z.submatrix_column, Z.descriptor, &work[0], &lwork, &info);
+      if (all_eigenpairs)
+        {
+          psyev(&jobz, &uplo, &n_rows, A_loc, &submatrix_row, &submatrix_column, descriptor, &ev[0],
+                eigenvectors_loc, &eigenvectors->submatrix_row, &eigenvectors->submatrix_column, eigenvectors->descriptor,
+                &work[0], &lwork, &info);
+
+          AssertThrow (info==0, LAPACKSupport::ExcErrorCode("psyev", info));
+        }
+      else
+        {
+          liwork = iwork[0];
+          AssertThrow(liwork>0,ExcInternalError());
+          iwork.resize(liwork);
 
-      AssertThrow (info==0, LAPACKSupport::ExcErrorCode("psyev", info));
+          psyevx(&jobz, &range, &uplo, &n_rows, A_loc, &submatrix_row, &submatrix_column, descriptor,
+                 &vl, &vu, &il, &iu, &abstol, &m, &nz, &ev[0], &orfac,
+                 eigenvectors_loc, &eigenvectors->submatrix_row, &eigenvectors->submatrix_column, eigenvectors->descriptor,
+                 &work[0], &lwork, &iwork[0], &liwork, &ifail[0], &iclustr[0], &gap[0], &info);
+
+          AssertThrow (info==0, LAPACKSupport::ExcErrorCode("psyevx", info));
+        }
+      // if eigenvectors are queried copy eigenvectors to original matrix
+      // as the temporary matrix eigenvectors has identical dimensions and
+      // block-cyclic distribution we simply swap the local array
+      if (compute_eigenvectors)
+        this->values.swap(eigenvectors->values);
+
+      //adapt the size of ev to fit m upon return
+      while ((int)ev.size() > m)
+        ev.pop_back();
     }
+  /*
+   * send number of computed eigenvalues to inactive processes
+   */
+  grid->send_to_inactive(&m, 1);
+
+  /*
+   * inactive processes have to resize array of eigenvalues
+   */
+  if (! grid->mpi_process_is_active)
+    ev.resize (m);
   /*
    * send the eigenvalues to processors not being part of the process grid
    */
   grid->send_to_inactive(ev.data(), ev.size());
 
   /*
-   * On exit, the lower triangle (if uplo='L') or the upper triangle (if uplo='U') of A,
-   * including the diagonal, is destroyed. Therefore, the matrix is unusable
+   * if only eigenvalues are queried the content of the matrix will be destroyed
+   * if the eigenpairs are queried matrix A on exit stores the eigenvectors in the columns
    */
-  state = LAPACKSupport::unusable;
+  property = LAPACKSupport::Property::general;
+  state = LAPACKSupport::eigenvalues;
```

Comment:
judging from comment above, should not there by `if..else` here?

Reply:
Yes!